addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/handler/MiHomeSubdevicesHandler.java

```diff
@@ -0,0 +1,651 @@
+import org.eclipse.smarthome.core.thing.ThingStatusInfo;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.mihome.MiHomeBindingConstants;
+import org.openhab.binding.mihome.internal.api.constants.DeviceConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Sets;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+/**
+ * Handler for the Mi|Home Subdevices
+ *
+ * @author Svilen Valkanov - Initial contribution
+ */
+public class MiHomeSubdevicesHandler extends BaseThingHandler {
+
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Sets.newHashSet(THING_TYPE_ENERGY_MONITOR,
+            THING_TYPE_MOTION_SENSOR, THING_TYPE_OPEN_SENSOR);
+
+    /**
+     * The number of pairing retries that the handler will make
+     */
+    private static final int PAIRING_RETRY_COUNT = 3;
+
+    /**
+     * Time in milliseconds between a single pairing start and end.
+     * The user has to press the pairing button in this interval.
+     */
+    private static final int WAIT_TIME_PAIRING = 15000;
+
+    /**
+     * Default update interval in seconds
+     */
+    public static final BigDecimal DEFAULT_UPDATE_INTERVAL = new BigDecimal(10);
+
+    private Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    /**
+     * Update interval in seconds
+     */
+    private long updateInterval;
+
+    /**
+     * A unique device representation in the MiHome REST API. Required when executing most of the requests to the
+     * Mi|Home REST API
+     */
+    private Integer mihomeID;
+
+    private ScheduledFuture<?> updateTask;
+
+    private ScheduledFuture<?> pairingTask;
+
+    private MiHomeGatewayHandler gatewayHandler;
+
+    public MiHomeSubdevicesHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing thing with UID {}", getThing().getUID().getAsString());
+
+        // The configuration is validated by the framework before calling initialize() on the handler
+        Configuration configuration = getConfig();
+        BigDecimal interval = (BigDecimal) configuration.get(MiHomeBindingConstants.CONFIG_UPDATE_ITNERVAL);
+        this.updateInterval = interval.longValue();
+
+        Bridge bridge = getBridge();
+        if (bridge != null) {
+            // Set the gateway handler
+            ThingHandler handler = bridge.getHandler();
+            if (handler instanceof MiHomeGatewayHandler) {
```

Comment:
The type should be always MiHomeGatewayHandler, I just didn't want to cast without calling instanceof first.